File: go/dlt645/transport/client/rtu_client.go
package client

import (
	"errors"
	"fmt"
	"log"
	"main/dlt645/common"
	"time"

	"github.com/tarm/serial"
)

type RtuClient struct {
	Port     string
	BaudRate int
	DataBits int
	StopBits int
	Parity   serial.Parity
	Timeout  time.Duration
	conn     *serial.Port
}

func (c *RtuClient) Configure(params interface{}) error {
	if params == nil {
		return errors.New("configuration params is nil")
	}
	// 类型断言，确保 params 是 RtuClient 类型的指针
	config, ok := params.(*RtuClient)
	if !ok {
		return errors.New("invalid configuration type, expected *RtuClient")
	}

	// 赋值配置参数
	c.Port = config.Port
	c.DataBits = config.DataBits
	c.StopBits = config.StopBits
	c.BaudRate = config.BaudRate
	c.Parity = config.Parity
	c.Timeout = config.Timeout
	log.Printf("RTU客户端配置成功: 端口=%s, 波特率=%d, 数据位=%d, 停止位=%d, 校验=%v, 超时=%v",
		c.Port, c.BaudRate, c.DataBits, c.StopBits, c.Parity, c.Timeout)
	return nil
}

func (c *RtuClient) Connect() error {
	config := &serial.Config{
		Name:        c.Port,
		Parity:      c.Parity,
		Baud:        c.BaudRate,
		Size:        byte(c.DataBits),
		StopBits:    serial.StopBits(byte(c.StopBits)),
		ReadTimeout: c.Timeout,
	}

	conn, err := serial.OpenPort(config)
	if err != nil {
		return fmt.Errorf("failed to open serial port: %w", err)
	}
	c.conn = conn

	log.Printf("Rtu client connected to port %s", c.Port)
	return nil
}

func (c *RtuClient) Disconnect() error {
	if c.conn == nil {
		return nil
	}
	err := c.conn.Close()
	if err != nil {
		return fmt.Errorf("failed to close serial port: %w", err)
	}
	c.conn = nil
	log.Printf("Rtu client disconnected from port %s", c.Port)
	return nil
}

// SendRequest 发送请求并接收响应
func (c *RtuClient) SendRequest(data []byte) ([]byte, error) {
	if c.conn == nil {
		return nil, errors.New("未连接到串口")
	}

	// 发送请求
	written, err := c.conn.Write(data)
	if err != nil {
		return nil, fmt.Errorf("发送请求失败: %v", err)
	}
	log.Printf("已发送请求: %v (%d bytes)", common.BytesToSpacedHex(data), written)

	// 使用动态缓冲区接收完整的响应帧
	var response []byte
	buf := make([]byte, 256)
	startTime := time.Now()
	found := false

	// 循环读取直到找到完整的帧或超时
	for time.Since(startTime) < c.Timeout && !found {
		n, err := c.conn.Read(buf)
		if err != nil {
			// 如果是超时但已经读取了部分数据，继续等待
			if n > 0 && time.Since(startTime) < c.Timeout {
				log.Printf("部分读取: %v (%d bytes)", common.BytesToSpacedHex(buf[:n]), n)
				response = append(response, buf[:n]...)
				continue
			}
			return nil, fmt.Errorf("接收响应失败: %v", err)
		}

		if n > 0 {
			response = append(response, buf[:n]...)
			log.Printf("读取数据: %v (%d bytes)", common.BytesToSpacedHex(buf[:n]), n)

			// 检查是否是有效的DLT645帧
			if len(response) >= 11 { // 最小帧长度
				// 检查帧结束标记 (DLT645协议通常以0x16结尾)
				for i := range response {
					if i >= 10 && response[i] == 0x16 {
						// 检查是否有足够的数据构成完整帧
						if i+1 >= 11 {
							found = true
							break
						}
					}
				}
			}
		}
	}

	if !found && len(response) == 0 {
		return nil, errors.New("未接收到任何响应数据")
	}

	log.Printf("已接收完整响应: %v (%d bytes)", common.BytesToSpacedHex(response), len(response))
	return response, nil
}

File: python/src/transport/client/rtu_client.py
"""同步 DL/T 645 串口客户端。"""

import threading
import time
from typing import Any, Optional

import serial

from ...common.message_capture import MessageCapture
from ...common.transform import bytes_to_spaced_hex
from ...protocol.protocol import DLT645Protocol
from ...transport.client.log import log


class RtuClient:
    """支持分片读取、总超时与重试的串口客户端。"""

    MAX_BUFFER_SIZE = 4096

    def __init__(
        self,
        port: str = "",
        baud_rate: int = 9600,
        data_bits: int = 8,
        stop_bits: int = 1,
        parity: str = serial.PARITY_NONE,
        timeout: float = 1.0,
    ) -> None:
        self.port = port
        self.baud_rate = baud_rate
        self.data_bits = data_bits
        self.stop_bits = stop_bits
        self.parity = parity
        self.timeout = timeout
        self.conn: Optional[serial.SerialBase] = None
        self._request_lock = threading.Lock()
        self._read_buffer = bytearray()
        self._message_capture: Optional[MessageCapture] = None

    def connect(self) -> bool:
        if self.conn is not None and self.conn.is_open:
            return True
        try:
            self.conn = serial.serial_for_url(
                self.port,
                baudrate=self.baud_rate,
                bytesize=self.data_bits,
                stopbits=self.stop_bits,
                parity=self.parity,
                timeout=self.timeout,
                write_timeout=self.timeout,
            )
            self._read_buffer.clear()
            log.info(f"RTU client connected to {self.port}")
            return True
        except (OSError, serial.SerialException, ValueError) as exc:
            self.conn = None
            log.error(f"Failed to open serial port {self.port}: {exc}")
            return False

    def disconnect(self) -> bool:
        conn, self.conn = self.conn, None
        self._read_buffer.clear()
        if conn is None:
            return True
        try:
            conn.close()
            return True
        except (OSError, serial.SerialException) as exc:
            log.error(f"Failed to close serial port: {exc}")
            return False

    def _ensure_connection(self) -> bool:
        if self.conn is None or not self.conn.is_open:
            return self.connect()
        return True

    def _clear_input_buffer(self) -> None:
        self._read_buffer.clear()
        if self.conn is not None:
            self.conn.reset_input_buffer()

    def _take_complete_frame(self) -> Optional[bytes]:
        while self._read_buffer:
            original = bytes(self._read_buffer)
            remaining, frame = DLT645Protocol.deserialize_with_remaining(original)
            if frame is not None:
                consumed = len(original) - len(remaining)
                response = original[:consumed]
                self._read_buffer = bytearray(remaining)
                return response
            if remaining != original:
                self._read_buffer = bytearray(remaining)
                continue
            return None
        return None

    def send_request(self, data: bytes, retries: int = 1) -> Optional[bytes]:
        with self._request_lock:
            for attempt in range(retries + 1):
                if not self._ensure_connection():
                    if attempt < retries:
                        time.sleep(0.5 * (attempt + 1))
                    continue
                conn = self.conn
                assert conn is not None
                original_timeout = conn.timeout
                tx_id: Optional[str] = None
                try:
                    self._clear_input_buffer()
                    written = conn.write(data)
                    conn.flush()
                    if written != len(data):
                        raise OSError(f"incomplete serial write: {written}/{len(data)}")
                    log.info(f"TX: {bytes_to_spaced_hex(data)}")
                    if self._message_capture:
                        tx_id = self._message_capture.capture_tx(data)

                    deadline = time.monotonic() + self.timeout
                    while True:
                        response = self._take_complete_frame()
                        if response is not None:
                            log.info(f"RX: {bytes_to_spaced_hex(response)}")
                            if self._message_capture:
                                self._message_capture.capture_rx(response, tx_id)
                            return response

                        remaining_time = deadline - time.monotonic()
                        if remaining_time <= 0:
                            break
                        conn.timeout = remaining_time
                        chunk = conn.read(max(1, min(conn.in_waiting, 256)))
                        if chunk:
                            self._read_buffer.extend(chunk)
                            if len(self._read_buffer) > self.MAX_BUFFER_SIZE:
                                raise ValueError(
                                    "RTU receive buffer exceeded 4096 bytes"
                                )
                    log.warning("RTU response timed out or was incomplete")
                except (OSError, serial.SerialException, ValueError) as exc:
                    log.error(f"RTU request attempt {attempt + 1} failed: {exc}")
                finally:
                    if conn is not None and self.conn is conn:
                        setattr(conn, "timeout", original_timeout)

                if attempt < retries:
                    time.sleep(0.5 * (attempt + 1))
            return None

    def send_only(self, data: bytes) -> bool:
        with self._request_lock:
            if not self._ensure_connection():
                return False
            conn = self.conn
            assert conn is not None
            try:
                self._clear_input_buffer()
                written = conn.write(data)
                conn.flush()
                if written != len(data):
                    return False
                log.info(f"TX (no response expected): {bytes_to_spaced_hex(data)}")
                if self._message_capture:
                    self._message_capture.capture_tx(data)
                return True
            except (OSError, serial.SerialException) as exc:
                log.error(f"RTU send failed: {exc}")
                return False

    def __enter__(self) -> "RtuClient":
        if not self.connect():
            raise OSError(f"无法打开串口 {self.port}")
        return self

    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        self.disconnect()
